examples/java/src/main/java/org/apache/beam/examples/io/examplekafkaread/ReadFromKafka.java

```diff
@@ -0,0 +1,162 @@
+    }
+    ExampleKafkaReadOffsetEstimator offsetPoller =
+        new ExampleKafkaReadOffsetEstimator(
+            new KafkaConsumer<byte[], byte[]>(
+                ExampleKafkaReadIOUtils.getOffsetConsumerConfig(
+                    "tracker-" + topicPartition, consumerConfig)),
+            topicPartition);
+    return new GrowableOffsetRangeTracker(restriction.getFrom(), offsetPoller);
+  }
+
+  @GetInitialWatermarkEstimatorState
+  public Instant getInitialWatermarkEstimatorState(@Timestamp Instant currentElementTimestamp) {
+    return currentElementTimestamp;
+  }
+
+  /**
+   * For this example, we estimate the watermark based on the clock time we receive elements from
+   * Kafka. As such, we use a manual estimator that we update during ProcessElement.
+   */
+  @NewWatermarkEstimator
+  public WatermarkEstimator<Instant> newWatermarkEstimator(
+      @WatermarkEstimatorState Instant watermarkEstimatorState) {
+    return new Manual(watermarkEstimatorState);
+  }
+
+  /**
+   * @param topicPartition is required to know what topic this bundle is being consumed from
+   * @param tracker is required to claim records as we receive them
+   * @param watermarkEstimator is required to be updated as we receive records
+   * @param receiver for outputting the KV Pairs
+   */
+  @ProcessElement
+  public ProcessContinuation processElement(
+      @Element TopicPartition topicPartition,
+      RestrictionTracker<OffsetRange, Long> tracker,
+      WatermarkEstimator<Instant> watermarkEstimator,
+      OutputReceiver<KV<byte[], byte[]>> receiver) {
+    try (Consumer<byte[], byte[]> consumer = new KafkaConsumer<byte[], byte[]>(consumerConfig)) {
+      consumer.assign(ImmutableList.of(topicPartition));
+      long startOffset = tracker.currentRestriction().getFrom();
+      // Seek to the lowest element for the restriction before we poll
+      consumer.seek(topicPartition, startOffset);
+      ConsumerRecords<byte[], byte[]> rawRecords = ConsumerRecords.empty();
+
+      while (true) {
+        rawRecords = consumer.poll(KAFKA_POLL_TIMEOUT);
+        // When there are no records available for the current TopicPartition, self-checkpoint
+        // and move to process the next element.
+        if (rawRecords.isEmpty()) {
+          // If we receive no records, we still should update the watermark. This ensures that
+          // watermark driven logic continues to work downstream, even if the data on Kafka is
+          // sparse
+          ((ManualWatermarkEstimator<Instant>) watermarkEstimator).setWatermark(Instant.now());
+
+          // We continue processing, as there is still work to be done in the range of the tracker
+          return ProcessContinuation.resume();
+        }
+        for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+          if (!tracker.tryClaim(rawRecord.offset())) {
+            // Because we poll from Kafka, as opposed to querying for specific rows, we may receive
```

Comment:
> records that are beyond the scope of the tracker
Do we expect multiple workers to process the same subscription at the same time ? I thought a Kafka worker would own the full offset range for a given subscription.